File: sanmao-api/scripts/stop-local-tunnel.py
#!/usr/bin/env python3
from __future__ import annotations

import os
import subprocess
from pathlib import Path


def listener_pids(port: int) -> list[int]:
    result = subprocess.run(
        ['netstat', '-ano', '-p', 'tcp'],
        check=False,
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        return []
    pids: set[int] = set()
    for raw_line in result.stdout.splitlines():
        parts = raw_line.split()
        if len(parts) < 5 or parts[0].upper() != 'TCP':
            continue
        local_address = parts[1]
        state = parts[3].upper()
        try:
            pid = int(parts[4])
        except ValueError:
            continue
        if state != 'LISTENING' or pid <= 0:
            continue
        if local_address.endswith(f':{port}'):
            pids.add(pid)
    return list(pids)


def kill_pid(pid: int) -> None:
    if pid <= 0:
        return
    subprocess.run(['taskkill', '/PID', str(pid), '/F', '/T'], check=False, capture_output=True, text=True)


def main() -> int:
    listen_port = int(os.environ.get('SANMAO_TUNNEL_PORT', '13000'))
    pid_path = Path(os.environ.get('SANMAO_TUNNEL_PID', str(Path.home() / '.ssh' / 'sanmao-tunnel.pid')))

    try:
        pid_value = int(pid_path.read_text(encoding='utf-8').strip())
    except Exception:
        pid_value = 0

    if pid_value:
        kill_pid(pid_value)
        print(f'[tunnel] stopped pid={pid_value}')

    pid_path.unlink(missing_ok=True)

    for listener_pid in listener_pids(listen_port):
        kill_pid(listener_pid)
        print(f'[tunnel] removed listener pid={listener_pid}')

    return 0


if __name__ == '__main__':
    raise SystemExit(main())

File: sanmao-api/scripts/stop-local-tunnel.mjs
#!/usr/bin/env node
import fs from 'node:fs/promises';
import path from 'node:path';
import os from 'node:os';
import { spawnSync } from 'node:child_process';

function runCapture(command, args) {
  return spawnSync(command, args, {
    encoding: 'utf8',
    windowsHide: true,
    stdio: ['ignore', 'pipe', 'pipe'],
  });
}

function killPid(pid) {
  if (!pid) {
    return;
  }
  runCapture('taskkill', ['/PID', String(pid), '/F', '/T']);
}

function listenerPids(port) {
  const result = runCapture('netstat', ['-ano', '-p', 'tcp']);
  if (result.error || (result.status ?? 1) !== 0) {
    return [];
  }
  const pids = new Set();
  for (const line of result.stdout.split(/\r?\n/)) {
    const parts = line.trim().split(/\s+/);
    if (parts.length < 5 || parts[0].toUpperCase() !== 'TCP') {
      continue;
    }
    const localAddress = parts[1];
    const state = parts[3].toUpperCase();
    const pid = Number(parts[4]);
    if (state !== 'LISTENING' || !pid) {
      continue;
    }
    if (localAddress.endsWith(`:${port}`)) {
      pids.add(pid);
    }
  }
  return [...pids];
}

async function main() {
  const listenPort = Number(process.env.SANMAO_TUNNEL_PORT || '13000');
  const pidPath = process.env.SANMAO_TUNNEL_PID || path.join(os.homedir(), '.ssh', 'sanmao-tunnel.pid');

  try {
    const pid = Number((await fs.readFile(pidPath, 'utf8')).trim()) || 0;
    if (pid) {
      killPid(pid);
      console.log(`[tunnel] stopped pid=${pid}`);
    }
  } catch {
  }

  await fs.rm(pidPath, { force: true });

  for (const pid of listenerPids(listenPort)) {
    killPid(pid);
    console.log(`[tunnel] removed listener pid=${pid}`);
  }
}

main().catch((error) => {
  console.error(error.message);
  process.exit(1);
});
